### ISSUE TITLE
Calling geojson.updateData multiple times in a short period of time can cause data loss

### ISSUE THREAD

**Issue by just-ads** [mapbox/mapbox-gl-js#13198]
Especially when deleting.

I have three features to be deleted;

Now I call `geojson.updateData` 3 times in a row, the second one may not be deleted.

![image](https://github.com/mapbox/mapbox-gl-js/assets/50706050/a772cc92-4467-4d0c-ac00-087ec4abb63f)

I checked the source code and found that `updateData` directly overwrites `_data`. If the previous `_updateWorkerData` is not completed, will some of the data that should be changed be lost if it is called multiple times in succession?

**Comment by mourner** [comment]
Yeah, good catch — we should probably coalesce multiple updates into one if the second one comes in while the first one wasn't yet sent to the worker.

**Comment by just-ads** [comment]
> Yeah, good catch — we should probably coalesce multiple updates into one if the second one comes in while the first one wasn't yet sent to the worker.
Right，i currently circumvent this bug by using FeatureCollection, but in this way I have to collect the features to be deleted in different functions together before deleting them.

**Comment by mourner** [comment]
This should be fixed in the upcoming v3.5.0-beta.1.

**Comment by just-ads** [comment]
> @mourner as this topic is hot right now, there is an issue with `source.setData` too #11953 . It's possible to call `setData` so fast, that there will be a visible desync in what was really set vs what is visible on map. In our use case it's visible in a way where user hovers over some interactive element to show line and symbols (icons) on map but quickly moves mouse out of this element, which results in quick call `setData( some data )` and immediately after `setData([])` - for whatever reason, symbol remains for a while, line always disappears.
> 
> In my issue, top right cat should disappear after 999 calls of `setData`, where in 1000th call im setting only 199 cats, however it disappears after some time. like a second or two.

This is because the timing returned by geojson.loadTile is incorrect.
First frame: Use `geojson.setData` to update a large amount of data
                 1-1: Trigger `loadTile` later
Second frame: Use `geojson.setData` to update to empty data
               2-1: Trigger `loadTile` later

At this time, 2-1 will complete the callback earlier than 1-1, which will eventually cause the data in the tile to be the data of the first frame.

The reason for this is that `geojson.loadTile` will eventually call `vector.loadTile`, and `workerTile.parse` in `vector.loadTile` will enter another worker in some cases, which may cause the next frame without this situation to complete the callback earlier than the previous frame.

Here is the code for `workerTile.parse`.
![image](https://github.com/user-attachments/assets/bd9ede6b-dc21-4c58-aa81-1d09d18bec53)

**Comment by just-ads** [comment]
@kamil-sienkiewicz-asi 
This is a temporary solution.
![image](https://github.com/user-attachments/assets/c68ae3ab-0d04-48e6-87c3-03cab2cfa1e2)
Just filter out callbacks that are less than the last request time

**Comment by kamil-sienkiewicz-asi** [comment]
@stepankuzmin Hey, is this something that will have priority to be fixed in near future? Seems like @just-ads found an issue in logic.
